### PR TITLE
Bugfix: GUI: Recognise NETWORK_LIMITED in formatServicesStr

### PR DESCRIPTION
**Pull request by luke-jr**
Currently, only the bottom 8 service bits are shown in the GUI peer details view.

`NODE_NETWORK_LIMITED` is the 11th bit (2^10).

The first commit expands the range to cover the full 64 bits, and properly label `"NETWORK_LIMITED"`.
The second commit refactors the code so that any future omitted service bits will trigger a compile warning.

### CODE REVIEW

File: src/protocol.h [hunk 1]
```diff
@@ -245,6 +245,7 @@ const std::vector<std::string> &getAllNetMessageTypes();
 
 /** nServices flags */
 enum ServiceFlags : uint64_t {
+    // NOTE: When adding here, be sure to update qt/guiutil.cpp's formatServicesStr too
```
Comment (hebasto) [inline src/protocol.h]:
```suggestion
    // NOTE: When adding here, be sure to update qt/guiutil.cpp's serviceFlagToStr too
```

File: src/qt/guiutil.cpp [hunk 2]
```diff
@@ -815,6 +815,24 @@ QString formatDurationStr(int secs)
...
+{
+    switch (ServiceFlags(mask)) {
+    case NODE_NONE: abort();  // impossible
+    case NODE_NETWORK:         return "NETWORK";
+    case NODE_GETUTXO:         return "GETUTXO";
+    case NODE_BLOOM:           return "BLOOM";
+    case NODE_WITNESS:         return "WITNESS";
+    case NODE_NETWORK_LIMITED: return "NETWORK_LIMITED";
+    // Not using default, so we get warned when a case is missing
+    }
```
Comment (hebasto) [inline src/qt/guiutil.cpp]: Could it follow [Developer Notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#c-data-structures)?

```suggestion
    } // no default case, so the compiler can warn about missing cases
```

**Review by hebasto — CHANGES_REQUESTED** [review]
Concept ACK 4341bffb6ef10909f3721329db27c5dc9bc720dd

File: src/qt/guiutil.cpp [hunk 3]
```diff
@@ -815,32 +815,33 @@ QString formatDurationStr(int secs)
...
+
 QString formatServicesStr(quint64 mask)
 {
     QStringList strList;
 
-    // Just scan the last 8 bits for now.
-    for (int i = 0; i < 8; i++) {
-        uint64_t check = 1 << i;
+    for (int i = 0; i < 64; i++) {
+        uint64_t check = 1LL << i;
```
Comment (Empact) [inline src/qt/guiutil.cpp]: nit: Could do the following, which may be more straightforward:

```cpp
for (uint64_t check = 1; check; check = check << 1) {
```

File: src/qt/guiutil.cpp [hunk 4]
```diff
@@ -815,32 +815,33 @@ QString formatDurationStr(int secs)
...
+    case NODE_GETUTXO:         return "GETUTXO";
+    case NODE_BLOOM:           return "BLOOM";
+    case NODE_WITNESS:         return "WITNESS";
+    case NODE_NETWORK_LIMITED: return "NETWORK_LIMITED";
+    // Not using default, so we get warned when a case is missing
+    }
+    if (bit < 8) {
+        return QString("%1[%2]").arg("UNKNOWN").arg(mask);
+    } else {
+        return QString("%1[2^%2]").arg("UNKNOWN").arg(bit);
```
Comment (laanwj) [inline src/qt/guiutil.cpp]: Why do bits `>=8` get `2^` but `<8` don't?

Comment (luke-jr) [reply src/qt/guiutil.cpp]: To use the same values that were shown previously for such bits. I'm assuming eventually we'll have all 8 assigned to literal strings, and the special case can go away then.

File: src/qt/guiutil.cpp (hunk 3 above)
Comment (luke-jr) [reply src/qt/guiutil.cpp]: Then we don't have the bit number, and would need to calculate it (which is more complex than the inverse).

File: src/qt/guiutil.cpp (hunk 4 above)
Comment (luke-jr) [reply src/qt/guiutil.cpp]: (Also because 9223372036854775808 is a pretty big number)

Comment (laanwj) [reply src/qt/guiutil.cpp]: OH! I get it now.
But personally I'd show 2^X for all bits then from now on, there's no need for complete version to version consistency in the GUI.

**Comment by jonasschnelli** [comment]
utACK 4341bffb6ef10909f3721329db27c5dc9bc720dd

**Comment by jonasschnelli** [comment]
Tested ACK 4341bffb6ef10909f3721329db27c5dc9bc720dd

Bildschirmfoto 2019-11-15 um 12 59 29

**Pull request merged**